Extend LU decomposition to handle non-square and singular matrices

twhunt opened the pull request:
Extends LUdecomposition() and LUdecomposition_Simple() to handle both singular, and non-square matrices.

Addresses issue #9796.
Supersedes pull request #9797 (I realized after the fact that deleting the fork associated with my previous pull request is not the preferred method, sorry). 
Labeled WIP since the test at line 1835 sympy/matrices/tests/test_matrices.py fails. I'd appreciate any insight about what this particular test is testing.

In the current implementations, ` sympy.Matrix([[1,2,3],[4,5,6],[7,8,9]]).LUdecomposition()` raises a ValueError, and `sympy.Matrix([[1,2],[4,5],[7,8]]).LUdecomposition()` raises a NonSquareMatrixError. However, every matrix has an LU decomposition when the rows are permuted appropriately.

The new implementations return:
(Matrix([
[1, 0, 0],
[4, 1, 0],
[7, 2, 1]]),
Matrix([
[1,  2,  3],
[0, -3, -6],
[0,  0,  0]]),
[])
and
(Matrix([
[1, 0, 0],
[4, 1, 0],
[7, 2, 1]]),
Matrix([
[1,  2],
[0, -3],
[0,  0]]),
[])

In this pull request, test_matrices.py no longer checks if LUdecomposition() raises an exception when the input is matrix non-square or singular matrix.

Added the following two matrices to test_LUdecomp():
Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9], [10, 11, 12]])
Matrix([[1, 2, 3], [2, 4, 6], [4, 5, 6]])

<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below.>
Fixes #9796

twhunt committed:
matrices: Extend LU decomposition to nonsquare and singular matrices

Extended LUdecomposition_Simple() and LUdecomposition() to handle both
singular, and non-square matrices.

Excised tests in test_matrices.py that checked if a non-square or
singular matrix raised an exception.
Also added the following two matrices to test_LUdecomp():
Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9], [10, 11, 12]])
Matrix([[1, 2, 3], [2, 4, 6], [4, 5, 6]])

Examples of new capabilities:
LU decomposition of a singular matrix:
A = sympy.Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
L, U, p = A.LUdecomposition()
sympy.pretty_print(A)
⎡1  2  3⎤
⎢       ⎥
⎢4  5  6⎥
⎢       ⎥
⎣7  8  9⎦
sympy.pretty_print(L)
⎡1  0  0⎤
⎢       ⎥
⎢4  1  0⎥
⎢       ⎥
⎣7  2  1⎦
sympy.pretty_print(U)
⎡1  2   3 ⎤
⎢         ⎥
⎢0  -3  -6⎥
⎢         ⎥
⎣0  0   0 ⎦
print p
[]

LU decomposition of a non-square matrix
(without the interactive interpreter input):
A=
⎡2   3   4 ⎤
⎢          ⎥
⎢9   12  15⎥
⎢          ⎥
⎢0   0   10⎥
⎢          ⎥
⎣16  29  35⎦

L=
⎡ 1     0      0    0⎤
⎢                    ⎥
⎢9/2    1      0    0⎥
⎢                    ⎥
⎢ 0     0      1    0⎥
⎢                    ⎥
⎣ 8   -10/3  -7/10  1⎦

U=
⎡2   3    4 ⎤
⎢           ⎥
⎢0  -3/2  -3⎥
⎢           ⎥
⎢0   0    10⎥
⎢           ⎥
⎣0   0    0 ⎦

p=
[[0, 1], [1, 2]]

twhunt committed:
Merge branch 'master' of https://github.com/sympy/sympy

twhunt committed:
Revert "matrices: Extend LU decomposition to nonsquare and singular matrices"

This reverts commit 78403147651632f0780d34555a16f8bbd0031783.

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -3912,35 +3950,87 @@ def LUdecomposition_Simple(self, iszerofunc=_iszero):
+
+        if self.rows == 0 or self.cols == 0:
+
+            raise ValueError('Matrix must have at least one entry to apply LUdecomposition_Simple().')
```
Could the method return an empty decomposition for an empty matrix instead of ValueError? That is what the current implementation does.

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -3912,35 +3950,87 @@ def LUdecomposition_Simple(self, iszerofunc=_iszero):
+            raise ValueError('Matrix must have at least one entry to apply LUdecomposition_Simple().')
+
+        lu = self.as_mutable()
+
```
Many blank lines make it harder to read the code. This is what PEP8 says:
 >   Use blank lines in functions, sparingly, to indicate logical sections.

twhunt committed:
matrices: LU decomposition of empty matrix doesn't raise error

When the input matrix has a zero length dimension,
LUdecomposition_simple() now returns a matrix of zeros of the same
size, and an empty list of row permutations.
The previous behavior was to raise a ValueError.

Example where number of columns in A is zero, and both
A.LUdecomposition_Simple() and A.LUdecomposition()
return an answer:
>>> A
Matrix(3, 0, [])
>>> lu, p = A.LUdecomposition_Simple()
>>> lu
Matrix(3, 0, [])
>>> p
[]
>>> L, U, p = A.LUdecomposition()
>>> L
Matrix([
[1, 0, 0],
[0, 1, 0],
[0, 0, 1]])
>>> U
Matrix(3, 0, [])
>>> p
[]
>>> L*U
Matrix(3, 0, [])

Cosmetic changes:
Reduced the number of blank lines down to a level that could
arguably be described as sparingly.
Replaced lu.shape[1] with lu.cols.

twhunt commented:
Followed @jksuom's suggestions, so that when A is m x n, where m or n is zero, then LUdecomposition_Simple() returns an m x n matrix instead of raising a ValueError.
Also removed some white space.

bjodah commented:
Just a remark: after tests are passing, one might want to run `http://github.com/sympy/sympy_benchmarks/` before merging to see how this branch compares with the previous implementation.

twhunt commented:
The test that fails is:
`raises(ValueError, lambda: Matrix([[5, 10, 7], [0, -1, 2], [8, 3, 4]]).LUdecomposition_Simple(iszerofunc=lambda x: abs(x) <= 4))`
I _believe_ that the purpose of this test is to verify that LUdecomposition_Simple raises a ValueError when every candidate pivot in the current pivot column is zero. Can someone more knowledgeable than me verify that this is the intent of the test? If so, then I would think it's safe to remove the test from the suite.

bjodah commented:
Perhaps raising a `ValueError` is the sane default for most people. We could add a keyword argument, e.g. `allow_singular=False` or something along those lines.

jksuom commented:
This is from the original commit 081e40b1:
> removed a redundant test for iszero and put a test in that tests a user-defined iszero function

So it looks like its purpose is to check that the `iszerofunc` mechanism is working. Perhaps it could be replaced with another test (not necessarily raising an error).

It also seems that the `iszerofunc` keyword is currently useless. It should be passed over to `LUdecomposition_Simple`.

twhunt committed:
matrices: Added pivot search strategy to LU decomposition functions

LUdecomposition_Simple() and LUdecomposition() now accept a keyword
argument that specifies a callable that returns the location of the
next pivot. The default behavior is to raise a ValueError if the
rank of the matrix being decomposed is strictly less than the number
of rows and columns. This mimics the behavior of the previous
implementation of LUdecomposition_Simple().
The new helper function _next_pivot_down() searches
only in the column below the current pivot location, and raises a
ValueError if no suitable pivot can be found.
The helper function _next_pivot_down_right() searches the entire
submatrix below and to the right of the current pivot position,
and signals that no pivot was found by returning the dimensions
of the matrix being decomposed.

Both LUdecomposition_Simple() and LUdecomposition() respect the
iszerofunc keyword argument.

twhunt committed:
matrices: New tests for iszerofunc LU decomposition keyword argument

Inside test_errors(), deleted the test:
raises(ValueError,
       lambda: Matrix([[5, 10, 7], [0, -1, 2], [8, 3, 4]]
       ).LUdecomposition_Simple(iszerofunc=lambda x: abs(x) <= 4))
which appeared to test if the keyword argument iszerofunc was being
used inside LUdecomposition_Simple().

Replaced with two standalone tests:
test_LUdecomposition_Simple_iszerofunc()
test_LUdecomposition_iszerofunc()
Both these tests work by defining a callable that is passed to
LUdecomposition_Simple() and LUdecomposition() that raises an
exception with a special error message.
The LU decomposition tester function  catches this error,
and tests the assertion that the error message matches the
expected special message.
If the LU decomposition function does not raise an error,
the test function fails with an assert False.

Passed non-default nextpivotfunc keyword argument to LU
decomposition functions in some	 tests.

twhunt commented:
I refactored LUdecomposition_Simple() so that its default behavior is to raise a ValueError if it tries to decompose a rank deficient matrix (rank < min(num_rows, num_cols). This mimics the behavior of the previous implementation.
Passing the keyword argument nextpivotfunc=_next_pivot_down_right causes LUdecomposition_Simple() to compute the LU decomposition of any matrix. An advantage of passing a next pivot callable instead of an is_singular boolean is increased flexibility. For example, the LDLT decomposition can be computed by LUdecomposition_Simple() by passing in a next pivot function that always returns the current pivot location. (D is the diagonal of U, assuming the input matrix is symmetric positive definite). 

I also added tests that verify that the iszerofunc keyword argument of LUdecomposition_Simple() and LUdecomposition() is respected, and deleted the previous test for this that relied on LUdecomposition_Simple raising a ValueErro.

twhunt committed:
test_matrices: Changed exception type used to communicate

In test_LUdecomposition_Simple_iszerofunc() and
test_LUdecomposition_iszerofunc(), changed the exception type
used to communicate from inside LUdecomposition() and
LUdecomsposition_Simple() to Exception. The previous exception type,
StandardException, is not in Python 3.3.
Communicates with args[0] member of Exception instance.

twhunt committed:
test_matrices: Changed exception type used to communicate

In test_LUdecomposition_Simple_iszerofunc() and
test_LUdecomposition_iszerofunc(), changed the exception type
used to communicate from inside LUdecomposition() and
LUdecomsposition_Simple() to Exception. The previous exception type,
StandardException, is not in Python 3.3.

twhunt committed:
Merge branch 'lu_decomp_extend' of https://github.com/twhunt/sympy into lu_decomp_extend

twhunt committed:
test_code_quality: Increased specificity of exception

sympy/utilities/tests/test_code_quality.py failed due to code
that raises an Exception, which the quality checker deemed
overly generic.
The code that tests if the LU decomposition functions respect
the iszerofunc keyword argument now raises a ValueError instead
of a function.

twhunt commented:
Do I need to do anything special since I altered some of the test modules in these commits? Thanks!

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -3912,35 +3974,65 @@ def LUdecomposition_Simple(self, iszerofunc=_iszero):
+                for col in range(0, pivot_row):
+                    tmp = lu[pivot_row, col]
+                    lu[pivot_row, col] = lu[pivot_row_cand, col]
+                    lu[pivot_row_cand, col] = tmp
```
It seems that these three lines could be replaced by a single one using python's swap idiom: `a, b = b, a`.

jksuom commented:
Would it be possible to work with only one nextpivotfunc, namely the more general one _next_pivot_down_right. Then there should be no need to add the keyword nextpivotfunc. The rank of the matrix could be checked at the end and ValueError could be returned if desired (maybe that could be controlled by a keyword).

I would also like to point out that SymPy coding follows the convention that there should be no blank lines after `def`, `if` and `for` lines.

Otherwise this looks good to me.

twhunt committed:
matrices: More idiomatic and whitespace tidying

Removed a temp variable used in a swap, and replaced
with a, b = b, a.
Removed blank lines.

twhunt commented:
I prefer to have some mechanism that allows the caller to specify the pivot strategy by passing a callable, although I'll happily defer to the judgement of any senior Sympy developers who object. I propose defining a new "experts only" function `_ludecomposition_simple()` that takes a keyword argument `nextpivotfunc`. `LUdecomposition_Simple()` would be a wrapper for `_ludecomposition_simple()`, and always pass it the more general `nextpivotfunc=_next_pivot_down_right`. `LUdecomposition_Simple()` would take a boolean keyword argument `decomposerankdeficient`, which determines if `LUdecomposition_Simple()` returns the LU factors or raises a ValueError if it is passed a matrix whose rank is strictly less than the number of its rows or columns. This would not break code that assumes the previous behavior of `LUdecomposition_Simple()`. It would also lay the groundwork for reusing code in the functions that rely on the  LDL decomposition and Gaussian Elimination functions in matrices.py, since the LU and LDL decompositions, and Gaussian elimination are all intimately related. What do you think @bjodah? Also, thanks again for your guidance so far.

bjodah commented:
Thanks for the ping. I'm afraid I'm busy during this week. I will try to find some time this weekend unless someone else find the time to review before then.

twhunt committed:
matrices: Added wrapper LU decomposition function

Moved the guts of LUdecomposition_simple() to LUCombined().
LUdecomposition_Simple() and LUdecomposition() now take a
keyword argument decomposerankdeficient, which determines if
the function raises a ValueError when the input matrix is rank
deficient, or returns the LU factors.
The default behavior is to raise a ValueError, which mimics
the behavior of the previous implementation.
LUCombined() computes the LU decomposition of its input,
regardless of rank.

Updated various tests in matrices/tests/test_matrices.py
use the decomposerankdeficient keyword argument instead of
nextpivotfunc argument.

twhunt commented:
I believe this pull request is now ready for review.

jksuom commented:
I think that the two pivoting strategies make the code unnecessarily complicated. There does not seem to be essential difference in the results.

bjodah commented:
I haven't had time to look into this. Maybe @siefkenj is interested and wants to share his view.

twhunt commented:
I agree that allowing multiple pivot strategies unavoidably increases complexity. However, the two pivot strategies do produce essentially different results. Given the singular matrix `A=Matrix([[1, 2, 3], [1, 2, 3], [1, 2, 3]])`, `LUdecomposition_Simple()` with the proposed pivot search produces an LU decomposition, while with the current pivot search, it raises a ValueError. As a consequence of the current pivot strategy,  `LUsolve()` cannot solve a linear system with infinitely many solutions. `gauss_jordan_solve()` can solve such a system, but it's worth noting that it relies on `rref()`, which computes the U factor from the LU decomposition as an intermediate step (under the proposed additional pivoting strategy). This suggests code overlap that could be removed, which would reduce the size of matrices.py, and arguably increase its maintainability. 

Here are a couple things I think argue in favor of keeping the ability to use the current pivot search algorithm. When passed a singular matrix, `det_LU_decomposition()` could return 0 as soon it encounters a zero pivot (although the current implementation raises a ValueError). Also, user code that depends on `LUdecomposition_Simple()` raising a ValueError when passed a rank deficient matrix would not break.

My two cents :)

jksuom commented:
It seems to me that the only difference between the two strategies is that ValueError may be raised in the middle of the computation if desired thus saving a little time. If there is only one strategy, then the computation would always be finished, but there would still be the possibility of raising the error if the rank turns out to be nonmaximal. A keyword could be used to determine whether the matrices are to be always returned or ValueError should be raised.

twhunt commented:
Fair enough. I'll refactor this weekend. Is there a sympy policy for maintaining backward compatibility, specifically regarding removing exceptions? I poked around the developer documentation, but couldn't find anything.

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -31,6 +31,33 @@ def _iszero(x):
+    dimensions of the input matrix."""
+    for candidate_pivot_col in range(pivot_col, matrix.cols):
+        for candidate_pivot_row in range(pivot_row, matrix.rows):
+            if not iszerofunc(matrix[candidate_pivot_row, candidate_pivot_col]):
```
If `iszerfunc` uses `True`, `False`, `None` logic, you may want to write this line as

```
if iszerofunc(matrix[candidate_pivot_row, candidate_pivot_col]) == False:
```

Look at https://github.com/sympy/sympy/blob/master/sympy/matrices/matrices.py#L5077 for a heuristic for finding a pivot.

bjodah commented:
I think it's fine stopping to raise exceptions when capabilities are added but maybe others disagree.

twhunt committed:
matrices: Revert LUdecomposition_Simple to internal pivot search

Refactored matrices.LUdecomposition_Simple() so that the pivot
search algorithm is defined inside the function, instead of calling
an external search function.
As a consequence, excised function _next_pivot_down_right(),
_next_pivot_down(), and LUCombined().
Removed import of _next_pivot_down_right from
matrices.tests.test_matrices.py.

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -3860,8 +3859,9 @@ def lower_triangular_solve(self, rhs):
-        """Returns the decomposition LU and the row swaps p.
+    def LUdecomposition(self, iszerofunc=_iszero, decomposerankdeficient=False):
+        """Returns the decomposition LU and the row swaps p as L, U, p.
+        See documentation for LUCombined for details about the keyword argument iszerofunc.
```
This long line could be split and a blank line added above.

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -3888,59 +3888,140 @@ def LUdecomposition(self, iszerofunc=_iszero):
-                        L[i, i] = 1
-                    U[i, j] = combined[i, j]
+
+        combined, p = self.LUdecomposition_Simple(iszerofunc=iszerofunc, decomposerankdeficient=decomposerankdeficient)
```
The argument list after ( could be moved to a new line.

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -3888,59 +3888,140 @@ def LUdecomposition(self, iszerofunc=_iszero):
+
+        # L is lower triangular self.rows x self.rows
+        # U is upper triangular self.rows x self.cols
+        # L has unit diagonal. For each column in combined, the subcolumn below the diagonal of combined is shared by L.
```
These long lines should also be split.

jksuom commented:
This looks good. Only, the code would be more readable in terminals if the line length were limited to 78 (mostly). The keyword `decomposerankdeficient` could perhaps be replaced by something less prone to typing errors. Maybe simple `check` (with inverted meaning) would suffice. I also agree with @bjodah  that its default could be `False`  (if I understand him correctly) unless that would break too many tests.

siefkenj commented:
`allowsingular` is slightly shorter than `decomposerankdeficient`.

twhunt committed:
matrices: LU decomposition default to decompose rank deficient

Changed default behavior to return LU decomposition of rank
deficient matrices rather than raise a ValueError.
Reduced maximum line length in LUdecomposition_simple() and
LUdecomposition().

twhunt commented:
I reduced the line lengths.  I also changed the default behavior to decompose rank deficient matrices, and all tests run by bin/test matrices pass under this change. As for changing decomposerankdeficient to allowsingular, the only definition I know of for a singular matrix requires that the matrix be square, which is not a requirement for the input of LUdecomposition_Simple(). I agree that decomposerankdeficient is somewhat clunky, but it does indicate what the keyword argument does. Any other suggestions?

siefkenj commented:
```
rankcheck
fullrankonly
ensurefullrank
allowrankdeficient
```

twhunt committed:
matrices: Change decomposerankdeficient keyword argument

Changed the keyword argument decomposerankdeficient to
allowrankdeficient.

twhunt commented:
`allowrankdeficient` it is!

siefkenj commented:
I thin `rankcheck` is a nice compromise.  Obviously, `rankcheck = False` would be equivalent to `allowrankdeficient = True`.  It's shorter, and it uses simpler words, which makes it better for non-native English speakers.

twhunt committed:
matrices: LU decomposition change allowrankdeficient to rankcheck

Changed the keyword argument of LUDecomposition_Simple()
and LUDecomposition() from allowrankdeficient, which defaulted
to True, to rankcheck, which defaults to false.
This is a name change only, and does not affect the behavior
of these two functions.

twhunt commented:
I'm not in 100% agreement about changing `allowrankdeficient` to `rankcheck` since the question that springs to _my_ mind is, "what does the rank check do?" That being said, given that I'm the only participant in this exchange that didn't give @siefkenj's comment a thumbs up, I'd go down in a crushing 3 to 1 defeat it is was put to a vote :) TL;DR, I changed `allowrankdeficient=True` to `rankcheck=False`.

siefkenj commented:
Would `checkrank` be more descriptive?  I personally think this option should be immediately deprecated.  Maybe I am missing something, but I cannot think of why you would want an LU decomposition routine that would bail on rank deficient matrices.  I think the fact that a unit test exists for this is someone being overzealous and it really should have been an `@XFAIL` instead of a test to verify the exception is thrown...

In summary, I hope whatever kwarg is added is slated for removal in sympy 1.2, so the name is perhaps not so important.

twhunt commented:
I wholeheartedly agree that any LU decomposition function should complete regardless of the rank of the matrix being decomposed. The caller knows best, and should handle the output accordingly. As a first time contributor, my main concern was to avoid breaking existing code that relied on the constrained behavior of the (hopefully) previous implementation of LUdecomposition_Simple(). If the keyword argument is to be deprecated or omitted, like I hope it will, then I really don't have a strong opinion about its name. I appreciate the feedback!

siefkenj commented:
@twhunt welcome to sympy!

You can look in `core/numbers.py` on line 784 for an example of raising a warning based on a function parameter.

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -3860,8 +3859,12 @@ def lower_triangular_solve(self, rhs):
-        """Returns the decomposition LU and the row swaps p.
+    def LUdecomposition(self,
+                        iszerofunc=_iszero,
+                        rankcheck=False):
```
Now that we're using the shorter `rankcheck`, I think this can fit on one line.

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -3860,8 +3859,12 @@ def lower_triangular_solve(self, rhs):
+    def LUdecomposition(self,
+                        iszerofunc=_iszero,
+                        rankcheck=False):
+        """Returns the decomposition LU and the row swaps p as L, U, p.
```
What about something more descriptive like

```
Returns (L,U,p) where L is a lower triangular matrix, U is an upper 
triangular matrix, and p is a list of row swaps.  If p is applied to `self`,
the result will be L*U.
```

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -3888,59 +3891,164 @@ def LUdecomposition(self, iszerofunc=_iszero):
+        # combined
+        for row_i in range(max_index):
+            for col_i in range(row_i, U.cols):
+                U[row_i, col_i] = combined[row_i, col_i]
```
You can do this in a much shorter way by using the fact that you can pass a function to `self._new`.  Look for places where `def entry(i, j)` is used in this file.

The basic idea would look like,
```
def entry_u(i,j):
    if i < j:
        return S.Zero
    return combined[i, j]

U = self._new(combined.rows, combined.cols, entry_u)
```

or something like that.  That way, you avoid using mutability to get the job done.  (If you want to do it your way, you need to do `self.zeros(...).as_mutable()` to get it to work with `ImmutableMatrix`.

twhunt replied:
I defined entry_U as:
```
def entry_U(i, j):
  return S.Zero if i > j else combined[i, j]
```

I looked in _new(), and it looks like the above ternary operator is evaluated for every entry index pair (i, j) when U is being constructed. Are there plans to introduce a mechanism where _new() accepts a sequence of row or column entry generators, and thus avoid self.rows*self.cols unnecessary index comparisons?

siefkenj replied:
No plans AFAIK.  I could see how that would be more efficient for sparse matrices, but how would it help for dense matrices?  Are you imagining something like

```
def row_gen(k):
    return [f(i) for i in range(k)] + [0]*(n-k)
```

?  How much faster is this than calling a function on every entry?  I'm not sure how you can generate entries and avoid a function call to every entry in general.  Nomatter what happens `simpify` is still called on every entry, so there is always `n*m` overhead with creating a matrix.

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -3888,59 +3891,164 @@ def LUdecomposition(self, iszerofunc=_iszero):
+        """Compute an lu decomposition of m x n matrix A, where P*A = L*U
+        L is m x m lower triangular with unit diagonal
+        U is m x n upper triangular
+        P is an m x m permutation matrix
```
Leave blank lines for new paragraphs.  This also might benefit from being a list.

For example
```
Compute an lu decomposition of m x n matrix A, where P*A = L*U

    * L is m x m lower triangular with unit diagonal
    * U is m x n upper triangular
    * P is an m x m permutation matrix

Returns an m x n ...
```

Sphinx will create nice HTML if you format this way.

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -3888,59 +3891,164 @@ def LUdecomposition(self, iszerofunc=_iszero):
+        The elements on the diagonal of L are all 1, and are not explicitly
+        stored.
+        U is stored in the upper triangular portion of lu, that is
+        lu[i ,j] = U[i, j] whenever i <= j.
```
Again, blank lines between paragraphs.  A diagram would also be really nice here.  Maybe something like?

```
Matrix([
[u, u, u, u],
[l, u, u, u],
[l, l, u, u],
[l, l, l, u]])
```

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -3888,59 +3891,164 @@ def LUdecomposition(self, iszerofunc=_iszero):
+        ValueError when passed a matrix whose rank is strictly less than
+        min(num rows, num cols). The default behavior is to decompose a rank
+        deficient matrix. Pass rankcheck=True to raise a
+        ValueError instead. (This mimics the previous behavior of this function).
```
Let's be bold and say `This behavior is deprecated since sympy 1.1`.

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -3888,59 +3891,164 @@ def LUdecomposition(self, iszerofunc=_iszero):
         ========
 
         LUdecomposition
+        LUdecomposition_Simple
```
No need to `See Also` yourself in your own docstring.

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -3888,59 +3891,164 @@ def LUdecomposition(self, iszerofunc=_iszero):
+                    if not iszeropivot:
+                        break
+                if not iszeropivot:
+                    break
```
You're doing a column by column search for a pivot, right?  I find the multiple negatives being used hard to follow.

There is a `_find_reasonable_pivot` function that you can use.  Maybe something like
```
for new_pivot_col in range(pivot_col, self.cols):
    pos, val, _, _ = _find_reasonable_pivot(lu[pivot_row:, new_pivot_col], iszerofunc)
    if pivot:
        new_pivot_row = pivot_row + pos
        break
```

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -3888,59 +3891,164 @@ def LUdecomposition(self, iszerofunc=_iszero):
+                # Two loops to handle case where pivot_col > pivot_row
+
+                for col in range(pivot_row + 1 if pivot_row == pivot_col
+                                 else pivot_col, lu.cols):
```
This is hard to parse with a big `if...else` clause in the argument list.  Can you save this to a well-named variable and then use it?

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -3888,59 +3891,164 @@ def LUdecomposition(self, iszerofunc=_iszero):
+                # Update L.
+                for col in range(0, pivot_row):
+                    lu[pivot_row, col], lu[candidate_pivot_row, col] =\
+                        lu[candidate_pivot_row, col], lu[pivot_row, col]
```
I haven't thought hard enough about what is going on here, but you might be able do write this using slices or the `.row_op()` method available to mutable matrices.

twhunt replied:
Shorthand for this operation is U[r, i:] = c*U[r_pivot, i:] + U[r, i:], where r > r_pivot.
Can .row_op() operate on only the elements to the right of some index for a row? There's probably some indexing trick I'm missing.

siefkenj replied:
Not that I know of, but `U[r, i:] = c*U[r_pivot, i:] + U[r, i:]` seems clearer than the loop to me.

siefkenj reviewed (changes requested):
You've done a great job updating some of the documentation, and I think this will be a nice change when it's fixed up.

twhunt commented:
@siefkenj Thanks for the detailed feedback! I'll integrate your requested changes ASAP, but I may not get to it until this weekend.

twhunt committed:
matrices: Documentation and LU implementation simplification

Added documentation to LUdecomposition() LUdecomposition_Simple().
Changed pivot search in LUdecomposition_Simple() to prefer
candidate pivots that iszerofunc can guarantee are zero over those
that iszerofunc returns None for.

Added deprecation warning when keyword argument  rankcheck is False.

twhunt committed:
matrices: Documentation and LU implementation simplification

Added documentation to LUdecomposition() LUdecomposition_Simple().
Changed pivot search in LUdecomposition_Simple() to prefer
candidate pivots that iszerofunc can guarantee are zero over those
that iszerofunc returns None for.

Added deprecation warning when keyword argument  rankcheck is False.

twhunt committed:
Merge branch 'lu_decomp_extend' of https://github.com/twhunt/sympy into lu_decomp_extend

twhunt committed:
matrices: Fixed LU decomposition early return bug

Fixed a bug where LUdecomposition_Simple() returned early when
iszeropivot was true after the pivot search loop completed.
The corrent test is:
if candidate_pivot_row is None and iszeropivot:
After the pivot search completes,  candidate_pivot_row is None
only when iszerofunc() returned true for every candidate pivot.

jksuom commented:
Is there something wrong with LUsolve that leads to an error in test_lagrange with these matrices.
```
(Pdb) p self
Matrix([
[1, 0, 0,        0,  0],
[0, 1, 0,        0,  0],
[0, 0, m,        0, -1],
[0, 0, 0, R**2*m/2,  R],
[0, 0, 1,       -R,  0]])
(Pdb) p rhs
Matrix([
[    Derivative(y(t), t)],
[Derivative(theta(t), t)],
[         g*m*sin(alpha)],
[                      0],
[                      0]])
```
The master code gives
```
(Pdb) self.LUsolve(rhs)
Matrix([
[    Derivative(y(t), t)],
[Derivative(theta(t), t)],
[       2*g*sin(alpha)/3],
[   2*g*sin(alpha)/(3*R)],
[      -g*m*sin(alpha)/3]])
```
while this PR returns
```
(Pdb) p self.LUsolve(rhs)
Matrix([
[    Derivative(y(t), t)],
[Derivative(theta(t), t)],
[         2*g*sin(alpha)],
[       2*g*sin(alpha)/R],
[        -g*m*sin(alpha)]])
```

twhunt commented:
@jksuom 
I need to augment the test suite since I introduced a bug that sympy/matrices/tests/test_matrices.py didn't catch. Fixing, and lesson learned :(

twhunt committed:
matrices: Fixed pivot factor bug, increased Pythonicity

Fixed a bug where the constant in a linear combination of
rows was incorrectly set to zero.

Removed explicit temporary tmp variable used in swap.

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -3888,22 +3895,78 @@ def LUdecomposition(self, iszerofunc=_iszero):
+        [u, u, u, u],
+        [l, u, u, u],
+        [l, l, u, u],
+        [l, l, l, u]])
```
This is great :-).  If you put it as an indented paragraph with blank lines before and after, it will render in the documentation as code instead of one line mushed together.

twhunt replied:
I had great inspiration ;)

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -3912,35 +3975,121 @@ def LUdecomposition_Simple(self, iszerofunc=_iszero):
+            candidate_pivot_row = None
+            while pivot_col != self.cols:
+                for r in range(pivot_row, self.rows):
+                    iszeropivot = iszerofunc(lu[r, pivot_col])
```
Is there a reason you didn't use the module-level function `_find_reasonable_pivot`?

twhunt replied:
I opted for the current implementation because _find_reasonable_pivot introduced a fairly significant slowdown for symbolic matrices, but I may be doing it wrong. To test _find_reasonable_pivot's performance, I added a call to it just before the existing pivot search in LUdecomposition_Simple(), and ignored its output:

```python
            while pivot_col != self.cols:
                for r in range(pivot_row, self.rows):

                    pivot_offset, pivot_val, assumed_nonzero, newly_determined =\
                        _find_reasonable_pivot(lu[pivot_row:, pivot_col],
                                               iszerofunc=iszerofunc,
                                               simpfunc=_simplify)

                    iszeropivot = iszerofunc(lu[r, pivot_col])
                    if iszeropivot is None and candidate_pivot_row is None:
                        # iszerofunc could not determine if
                        # lu[r, pivot_col] is zero.
                        # Save this row in case we don't
                        # find a nonzero entry during the search
                        # of this column
                        candidate_pivot_row = r
                    elif iszeropivot == False:
                        # iszerofunc determined that
                        # lu[r, pivot_col] is zero.
                        # Use this entry as the pivot
                        candidate_pivot_row = r
                        break
                    elif iszeropivot == True:
                        lu[r, pivot_col] = S.Zero
                if candidate_pivot_row is None:
                    # There is no nonzero pivot in this column.
                    # Exited search within column because
                    # iszeropivot() returned False for each
                    # entry
                    pivot_col += 1
                else:
                    # lu[r, pivot_col] is either non-zero,
                    # or iszeropivot could not determine that it is
                    # zero.
                    break
```
I then ran the following script twice with the call to _find_reasonable_pivot() commented out in LUdecomposition_Simple() on the second run.
```sympy
import sympy

def main():

    lu, perm = sympy.Matrix(4, 2, map(sympy.Symbol, "abcdefgh")).LUdecomposition_Simple()

    sympy.pretty_print(lu)
    print perm

if __name__ == "__main__":

    for k in range(10):
        main()
```
The profiler results of both runs are:
```
>>> p = pstats.Stats('/tmp/find_reasonable_pivot_on')
>>> p.sort_stats('cumulative').print_stats(10)
Wed Mar 29 21:06:21 2017    /tmp/find_reasonable_pivot_on

         21332950 function calls (20699591 primitive calls) in 55.268 seconds

   Ordered by: cumulative time
   List reduced from 2574 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   55.465   55.465 find_reasonable_pivot_test.py:1(<module>)
       10    0.000    0.000   53.589    5.359 find_reasonable_pivot_test.py:3(main)
       10    0.002    0.000   53.279    5.328 sympy/matrices/matrices.py:3930(LUdecomposition_Simple)
       70    0.005    0.000   53.149    0.759 sympy/matrices/matrices.py:5269(_find_reasonable_pivot)
  430/250    0.018    0.000   44.144    0.177 sympy/core/expr.py:595(equals)
       90    0.005    0.000   35.123    0.390 sympy/core/expr.py:440(is_constant)
      180    0.012    0.000   29.239    0.162 sympy/core/expr.py:362(_random)
      720    0.012    0.000   28.823    0.040 sympy/utilities/randtest.py:16(random_complex_number)
1800/1440    0.027    0.000   28.763    0.020 sympy/simplify/simplify.py:1159(nsimplify)
      360    0.029    0.000   28.669    0.080 sympy/simplify/simplify.py:1284(_real_to_rational)


<pstats.Stats instance at 0x10c6e9248>
>>> p = pstats.Stats('/tmp/find_reasonable_pivot_off')
>>> p.sort_stats('cumulative').print_stats(10)
Wed Mar 29 21:06:40 2017    /tmp/find_reasonable_pivot_off

         373170 function calls (366705 primitive calls) in 1.418 seconds

   Ordered by: cumulative time
   List reduced from 2108 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.001    0.001    1.424    1.424 find_reasonable_pivot_test.py:1(<module>)
        1    0.005    0.005    0.844    0.844 sympy/__init__.py:12(<module>)
       10    0.000    0.000    0.579    0.058 find_reasonable_pivot_test.py:3(main)
       10    0.000    0.000    0.448    0.045 sympy/printing/pretty/pretty.py:2162(pretty_print)
       10    0.000    0.000    0.430    0.043 sympy/printing/pretty/pretty.py:2144(pretty)
       10    0.000    0.000    0.429    0.043 sympy/printing/pretty/pretty.py:61(doprint)
   449/10    0.006    0.000    0.423    0.042 sympy/printing/printer.py:235(_print)
       10    0.000    0.000    0.423    0.042 sympy/printing/pretty/pretty.py:664(_print_MatrixBase)
       10    0.003    0.000    0.419    0.042 sympy/printing/pretty/pretty.py:595(_print_matrix_contents)
        1    0.005    0.005    0.331    0.331 sympy/polys/__init__.py:1(<module>)


<pstats.Stats instance at 0x10b011b48>
```

siefkenj replied:
You're right.  It is really slow on symbolic entries.  Unfortunately those cannot be proved to be non-zero, so `_find_reasonable_pivots` has to go through every level of check before it just gives up.  The danger of not using it is that the default `_iszerofunc` **will** fail for some terms which happen to be zero.  You could check the examles near the end of the tests for cases where this happens.  (I believe a matrix with entries the golden ratio and its powers fails).  

`_find_reasonable_pivots` should use the `_iszerofunc` that it is passed, and if `_iszerofunc` returns `False`, it should be quite quick.  Could you run the test again with `_iszerofunc = lambda x: True if x.is_zero is None else x.is_zero`  ?

siefkenj replied:
If your application requires the LU computation to be fast on symbolic input, you can leave it as is and reference bug #10279 . Put a comment

```
# XXX `_find_reasonable_pivot` uses slow zero testing. Blocked by bug #10279
```

above that line.

twhunt replied:
Looks like I was doing it wrong:
```
>>> p = pstats.Stats('/tmp/find_reasonable_pivot_on')
>>> p.sort_stats('cumulative').print_stats(10)
Thu Mar 30 18:43:46 2017    /tmp/find_reasonable_pivot_on

         384704 function calls (377602 primitive calls) in 2.781 seconds

   Ordered by: cumulative time
   List reduced from 2116 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.001    0.001    2.796    2.796 find_reasonable_pivot_test.py:1(<module>)
        1    0.005    0.005    1.983    1.983 sympy/__init__.py:12(<module>)
        1    0.026    0.026    0.885    0.885 sympy/polys/__init__.py:1(<module>)
       10    0.001    0.000    0.812    0.081 find_reasonable_pivot_test.py:3(main)
        1    0.005    0.005    0.669    0.669 sympy/core/__init__.py:2(<module>)
       10    0.000    0.000    0.573    0.057 sympy/printing/pretty/pretty.py:2162(pretty_print)
       10    0.058    0.006    0.538    0.054 sympy/printing/pretty/pretty.py:2144(pretty)
       10    0.000    0.000    0.480    0.048 sympy/printing/pretty/pretty.py:61(doprint)
   449/10    0.003    0.000    0.476    0.048 sympy/printing/printer.py:235(_print)
       10    0.000    0.000    0.476    0.048 sympy/printing/pretty/pretty.py:664(_print_MatrixBase)


<pstats.Stats instance at 0x10c447c68>
```

```
>>> p = pstats.Stats('/tmp/find_reasonable_pivot_off')
>>> p.sort_stats('cumulative').print_stats(10)
Thu Mar 30 18:44:09 2017    /tmp/find_reasonable_pivot_off

         373177 function calls (366716 primitive calls) in 2.467 seconds

   Ordered by: cumulative time
   List reduced from 2108 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    2.522    2.522 find_reasonable_pivot_test.py:1(<module>)
        1    0.027    0.027    1.900    1.900 sympy/__init__.py:12(<module>)
        1    0.003    0.003    0.670    0.670 sympy/core/__init__.py:2(<module>)
        1    0.014    0.014    0.628    0.628 sympy/polys/__init__.py:1(<module>)
       10    0.001    0.000    0.621    0.062 find_reasonable_pivot_test.py:3(main)
       10    0.000    0.000    0.426    0.043 sympy/printing/pretty/pretty.py:2162(pretty_print)
       10    0.000    0.000    0.395    0.039 sympy/printing/pretty/pretty.py:2144(pretty)
       10    0.000    0.000    0.394    0.039 sympy/printing/pretty/pretty.py:61(doprint)
   449/10    0.003    0.000    0.391    0.039 sympy/printing/printer.py:235(_print)
       10    0.000    0.000    0.391    0.039 sympy/printing/pretty/pretty.py:664(_print_MatrixBase)


<pstats.Stats instance at 0x10c299d40>
```
I'll base the pivot search method on `_find_reasonable_pivot()`. As long as the caller of `LUdecomposition_simple()` can turn off unwanted simplifications when it "knows" it doesn't need them, then I have no objection to using ` _find_reasonable_pivot()`.

twhunt replied:
Well this is embarrassing, I was calling _find_reasonable_limit in an inner loop, and not the correct outer loop. Those profile timings are meaningless.

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -3912,35 +3975,121 @@ def LUdecomposition_Simple(self, iszerofunc=_iszero):
+                # Store factors of L in the subcolumn below
+                # (pivot_row, pivot_row).
+                lu[row, pivot_row] =\
+                    lu[row, pivot_col]/lu[pivot_row, pivot_col]
```
space around `\`

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -3912,35 +3975,121 @@ def LUdecomposition_Simple(self, iszerofunc=_iszero):
+                                 " rankcheck=False to compute"
+                                 " the LU decomposition of this matrix."
+                                 " This behavior is deprecated since sympy"
+                                 " 1.1.")
```
I don't think the `ValueError` should include a deprecation warning aswell, especially since they've already been warned.

twhunt committed:
matrices: Use _find_reasonable_pivot() in LU decomposition

Changed the pivot search algorithm to use _find_reasonable_pivot().
Fixed IndexError bug in LUdecomposition() when L is built.
Tidied up some documentation fo LUdecomposition_Simple().

asmeurer commented:
Travis has a lot of timeouts. That suggests to me that this is slowing things down.

twhunt commented:
Leaving this here mostly as a note to myself:
Using `egrep '(>>>)|(^  \.\.\.)' doc/src/modules/physics/mechanics/examples/bicycle_example.rst | cut -c 7-` to extract the Python out of bicycle_example.rst produces:
```python
from sympy import *
from sympy.physics.mechanics import *
print('Calculation of Linearized Bicycle \"A\" Matrix, '
      'with States: Roll, Steer, Roll Rate, Steer Rate')
Vector.simp = False
mechanics_printing(pretty_print=False)
q1, q2, q3, q4, q5 = dynamicsymbols('q1 q2 q3 q4 q5')
q1d, q2d, q4d, q5d = dynamicsymbols('q1 q2 q4 q5', 1)
u1, u2, u3, u4, u5, u6 = dynamicsymbols('u1 u2 u3 u4 u5 u6')
u1d, u2d, u3d, u4d, u5d, u6d = dynamicsymbols('u1 u2 u3 u4 u5 u6', 1)
WFrad, WRrad, htangle, forkoffset = symbols('WFrad WRrad htangle forkoffset')
forklength, framelength, forkcg1 = symbols('forklength framelength forkcg1')
forkcg3, framecg1, framecg3, Iwr11 = symbols('forkcg3 framecg1 framecg3 Iwr11')
Iwr22, Iwf11, Iwf22, Iframe11 = symbols('Iwr22 Iwf11 Iwf22 Iframe11')
Iframe22, Iframe33, Iframe31, Ifork11 = \
    symbols('Iframe22 Iframe33 Iframe31 Ifork11')
Ifork22, Ifork33, Ifork31, g = symbols('Ifork22 Ifork33 Ifork31 g')
mframe, mfork, mwf, mwr = symbols('mframe mfork mwf mwr')
N = ReferenceFrame('N')
Y = N.orientnew('Y', 'Axis', [q1, N.z])
R = Y.orientnew('R', 'Axis', [q2, Y.x])
Frame = R.orientnew('Frame', 'Axis', [q4 + htangle, R.y])
WR = ReferenceFrame('WR')
TempFrame = Frame.orientnew('TempFrame', 'Axis', [-htangle, Frame.y])
Fork = Frame.orientnew('Fork', 'Axis', [q5, Frame.x])
TempFork = Fork.orientnew('TempFork', 'Axis', [-htangle, Fork.y])
WF = ReferenceFrame('WF')
WR_cont = Point('WR_cont')
WR_mc = WR_cont.locatenew('WR_mc', WRrad * R.z)
Steer = WR_mc.locatenew('Steer', framelength * Frame.z)
Frame_mc = WR_mc.locatenew('Frame_mc', -framecg1 * Frame.x + framecg3 * Frame.z)
Fork_mc = Steer.locatenew('Fork_mc', -forkcg1 * Fork.x + forkcg3 * Fork.z)
WF_mc = Steer.locatenew('WF_mc', forklength * Fork.x + forkoffset * Fork.z)
WF_cont = WF_mc.locatenew('WF_cont', WFrad*(dot(Fork.y, Y.z)*Fork.y - \
                                            Y.z).normalize())
Y.set_ang_vel(N, u1 * Y.z)
R.set_ang_vel(Y, u2 * R.x)
WR.set_ang_vel(Frame, u3 * Frame.y)
Frame.set_ang_vel(R, u4 * Frame.y)
Fork.set_ang_vel(Frame, u5 * Fork.x)
WF.set_ang_vel(Fork, u6 * Fork.y)
WR_cont.set_vel(N, 0)
WR_mc.v2pt_theory(WR_cont, N, WR)
Steer.v2pt_theory(WR_mc, N, Frame)
Frame_mc.v2pt_theory(WR_mc, N, Frame)
Fork_mc.v2pt_theory(Steer, N, Fork)
WF_mc.v2pt_theory(Steer, N, Fork)
WF_cont.v2pt_theory(WF_mc, N, WF)
Frame_I = (inertia(TempFrame, Iframe11, Iframe22, Iframe33, 0, 0,
                                                  Iframe31), Frame_mc)
Fork_I = (inertia(TempFork, Ifork11, Ifork22, Ifork33, 0, 0, Ifork31), Fork_mc)
WR_I = (inertia(Frame, Iwr11, Iwr22, Iwr11), WR_mc)
WF_I = (inertia(Fork, Iwf11, Iwf22, Iwf11), WF_mc)
BodyFrame = RigidBody('BodyFrame', Frame_mc, Frame, mframe, Frame_I)
BodyFork = RigidBody('BodyFork', Fork_mc, Fork, mfork, Fork_I)
BodyWR = RigidBody('BodyWR', WR_mc, WR, mwr, WR_I)
BodyWF = RigidBody('BodyWF', WF_mc, WF, mwf, WF_I)
print('Before Forming the List of Nonholonomic Constraints.')
kd = [q1d - u1, q2d - u2, q4d - u4, q5d - u5]
conlist_speed = [WF_cont.vel(N) & Y.x,
                 WF_cont.vel(N) & Y.y,
                 WF_cont.vel(N) & Y.z]
conlist_coord = [WF_cont.pos_from(WR_cont) & Y.z]
FL = [(Frame_mc, -mframe * g * Y.z), (Fork_mc, -mfork * g * Y.z),
      (WF_mc, -mwf * g * Y.z), (WR_mc, -mwr * g * Y.z)]
BL = [BodyFrame, BodyFork, BodyWR, BodyWF]
KM = KanesMethod(N, q_ind=[q1, q2, q5],
          q_dependent=[q4], configuration_constraints=conlist_coord,
          u_ind=[u2, u3, u5],
          u_dependent=[u1, u4, u6], velocity_constraints=conlist_speed,
          kd_eqs=kd)
print('Before Forming Generalized Active and Inertia Forces, Fr and Fr*')
(fr, frstar) = KM.kanes_equations(BL, FL)
print('Base Equations of Motion Computed')
PaperRadRear  =  0.3
PaperRadFront =  0.35
HTA           =  evalf.N(pi/2-pi/10)
TrailPaper    =  0.08
rake          =  evalf.N(-(TrailPaper*sin(HTA)-(PaperRadFront*cos(HTA))))
PaperWb       =  1.02
PaperFrameCgX =  0.3
PaperFrameCgZ =  0.9
PaperForkCgX  =  0.9
PaperForkCgZ  =  0.7
FrameLength   =  evalf.N(PaperWb*sin(HTA) - (rake - \
                        (PaperRadFront - PaperRadRear)*cos(HTA)))
FrameCGNorm   =  evalf.N((PaperFrameCgZ - PaperRadRear - \
                         (PaperFrameCgX/sin(HTA))*cos(HTA))*sin(HTA))
FrameCGPar    =  evalf.N((PaperFrameCgX / sin(HTA) + \
                         (PaperFrameCgZ - PaperRadRear - \
                          PaperFrameCgX / sin(HTA) * cos(HTA)) * cos(HTA)))
tempa         =  evalf.N((PaperForkCgZ - PaperRadFront))
tempb         =  evalf.N((PaperWb-PaperForkCgX))
tempc         =  evalf.N(sqrt(tempa**2 + tempb**2))
PaperForkL    =  evalf.N((PaperWb*cos(HTA) - \
                         (PaperRadFront - PaperRadRear)*sin(HTA)))
ForkCGNorm    =  evalf.N(rake + (tempc * sin(pi/2 - \
                         HTA - acos(tempa/tempc))))
ForkCGPar     =  evalf.N(tempc * cos((pi/2 - HTA) - \
                         acos(tempa/tempc)) - PaperForkL)
v = Symbol('v')
val_dict = {
      WFrad: PaperRadFront,
      WRrad: PaperRadRear,
      htangle: HTA,
      forkoffset: rake,
      forklength: PaperForkL,
      framelength: FrameLength,
      forkcg1: ForkCGPar,
      forkcg3: ForkCGNorm,
      framecg1: FrameCGNorm,
      framecg3: FrameCGPar,
      Iwr11: 0.0603,
      Iwr22: 0.12,
      Iwf11: 0.1405,
      Iwf22: 0.28,
      Ifork11: 0.05892,
      Ifork22: 0.06,
      Ifork33: 0.00708,
      Ifork31: 0.00756,
      Iframe11: 9.2,
      Iframe22: 11,
      Iframe33: 2.8,
      Iframe31: -2.4,
      mfork: 4,
      mframe: 85,
      mwf: 3,
      mwr: 2,
      g: 9.81,
      q1: 0,
      q2: 0,
      q4: 0,
      q5: 0,
      u1: 0,
      u2: 0,
      u3: v/PaperRadRear,
      u4: 0,
      u5: 0,
      u6: v/PaperRadFront}
kdd = KM.kindiffdict()
print('Before Linearization of the \"Forcing\" Term')
# forcing_lin = KM.linearize()[0].subs(sub_dict)
MM_full = (KM._k_kqdot).row_join(zeros(4, 6)).col_join(
          (zeros(6, 4)).row_join(KM.mass_matrix))
print('Before Substitution of Numerical Values')
# MM_full = MM_full.subs(val_dict)
# forcing_lin = forcing_lin.subs(val_dict)
# print('Before .evalf() call')
# MM_full = MM_full.evalf()
# forcing_lin = forcing_lin.evalf()
# Amat = MM_full.inv() * forcing_lin
# A = Amat.extract([1,2,4,6],[1,2,3,5])
# print(A)
# print('v = 1')
# print(A.subs(v, 1).eigenvals())
# print('v = 2')
# print(A.subs(v, 2).eigenvals())
# print('v = 3')
# print(A.subs(v, 3).eigenvals())
# print('v = 4')
# print(A.subs(v, 4).eigenvals())
# print('v = 5')
# print(A.subs(v, 5).eigenvals())
```
This script completes within twenty seconds when I run it locally, whereas `bin/doctest doc/src/modules/physics/mechanics/examples/bicycle_example.rst` runs for hours without completing.

twhunt committed:
matrices: Added simpfunc keyword argument to LU decomposition

Added simpfunc keyword argument to LUdecomposition_Simple()
and LUdecomposition(). simpfunc is used by the pivot search
when iszerofunc cannot guarantee that a pivot candidate is
nonzero.

twhunt committed:
matrices: LU partial reversion to naive pivot search strategy

The doctests timeout when the LU decomposition functions use
matrices._find_reasonable_pivot() as the pivot search.
Future versions of the LU decomposition functions can pass
the keyword arguments iszerofunc and simpfunc to a future
versionof _find_reasonable_pivot() that completes in less
time.
The current pivot search algorithm employed by
LUdecomposition_simple() first looks for a pivot that iszerofunc
guarantees is nonzero. If no such candidate exists,
the candidates are first simplified if the caller does not pass
simpfunc=None, and then a second pivot search more permissive pivot
search is employed. In the second search, the pivot is the first
candidate that iszerofunc can guarantee is nonzero.
If no such candidate pivot exists, then the pivot is the first
candidate for which iszerofunc returns None.
If no such candidate exists, the pivot search starts again in the
column to the right.

twhunt committed:
Merge branch 'master' into lu_decomp_extend

twhunt commented:
Reverted to a partially naive pivot search, since the unit tests time timeout when the the pivot search uses `_find_reasonable_pivot()`. By default, the current pivot search algorithm does not simplify any of the candidate pivots, although the caller can pass a simplification function via the `simpfunc` keyword argument. Looking towards the future, the LU decomposition functions take keyword arguments `simpfunc` and `iszerofunc`, which can be passed to a future version of `_find_reasonable_pivot()` that completes in less time.

siefkenj commented:
@twhunt Would you mind looking at PR #12521 ?  I created a new `_eval_rref` function that is much faster than the previous one.  I suspect it could be really useful for writing an improved LU decomposition function.

twhunt commented:
@siefkenj 
Looks intriguing, and I'll refactor ASAP. I probably won't get to it until mid week though.

twhunt committed:
Merge branch 'master' into lu_decomp_extend

twhunt committed:
matrices/tests/test_matrices.py:
Changed minorEntry to minor, and minorMatrix to minor_submatrix
so tests don't fail with deprecation warning.

twhunt commented:
@siefkenj 
Before I reverted to the naive pivot search, the The Travis tests timed out during a call to `_find_reasonable_pivot()`, and it looks like this function is unchanged in your commonmatrix5 branch. Is this correct? If so, then I don't see a way to integrate the more thorough and correct pivot search of `_find_reasonable_pivot()` into `LUdecompositionSimple()` without running into the same problem of having the Travis tests timeout.

siefkenj commented:
@twhunt `_find_reasonable_pivot()` is unchanged, but I also avoid doing any divisions until the very end, which speeds up the procedure quite a bit.  Can you leverage the `echelon_form` function added in PR #12521 do some of the work for the decomposition?  If so, I'd be curious if it gives the same slowdown.

If the slowdown cannot be fixed, then you should use a stub for `_find_reasonable_pivot()` that has the same return values but works faster, and mark it with a `#XXX workaround for bug ..." comment.

twhunt commented:
@siefkenj
If I understand the algorithm for zeroing below the pivot in your RREF algorithm, I don't see a straightforward way of leveraging it in the version of the LU decomposition that I'm familiar with. My thinking is:

To zero everything below the (0,0) position in v=
```
>>> sympy.pretty_print(v)
⎡a⎤
⎢ ⎥
⎢b⎥
⎢ ⎥
⎢c⎥
⎢ ⎥
⎣d⎦
```
the RREF code computes L_v*v. L_v, which is not explicitly constructed, is:
```
>>> sympy.pretty_print(L_v)
⎡1   0  0  0⎤
⎢           ⎥
⎢-b  a  0  0⎥
⎢           ⎥
⎢-c  0  a  0⎥
⎢           ⎥
⎣-d  0  0  a⎦
```
On the other hand, the LU decomposition code zeroes the entries below the pivot by performing LL_v*v, where LL_v is:
```
>>> sympy.pretty_print(LL_v)
⎡ 1   0  0  0⎤
⎢            ⎥
⎢-b          ⎥
⎢───  1  0  0⎥
⎢ a          ⎥
⎢            ⎥
⎢-c          ⎥
⎢───  0  1  0⎥
⎢ a          ⎥
⎢            ⎥
⎢-d          ⎥
⎢───  0  0  1⎥
⎣ a          ⎦
```
LL_v has two key properties that make computing the LU decomposition efficient. The first is that its inverse is computed by negating the nonzero entries below the diagonal:
```
>>> sympy.pretty_print(LL_v.inv())
⎡1  0  0  0⎤
⎢          ⎥
⎢b         ⎥
⎢─  1  0  0⎥
⎢a         ⎥
⎢          ⎥
⎢c         ⎥
⎢─  0  1  0⎥
⎢a         ⎥
⎢          ⎥
⎢d         ⎥
⎢─  0  0  1⎥
⎣a         ⎦
```
If LL_v0 zeros below the (0,0) position, LL_v1 zeroes below the (1,1) position, and so on, then the L factor in the LU decomposition is inv(LL_v0)\*inv(LL_v1)*...*inv(LL_vn).
The second nice property is that L can be formed by copying the entries below the (i,i) position of inv(LL_vi) into the appropriate location in L, instead of explicitly computing the product of inverses.

The inverse of a lower triangular factor from the RREF algorithm is a little more complicated:
```
>>> sympy.pretty_print(L_v.inv())
⎡1  0  0  0⎤
⎢          ⎥
⎢b  1      ⎥
⎢─  ─  0  0⎥
⎢a  a      ⎥
⎢          ⎥
⎢c     1   ⎥
⎢─  0  ─  0⎥
⎢a     a   ⎥
⎢          ⎥
⎢d        1⎥
⎢─  0  0  ─⎥
⎣a        a⎦
```
In the RREF algorithm, one could encode the sequence of row operations that transform A to lower triangular form by computing inv(L_v0)\*inv(L_v1)\*...\*inv(L_vn). However, I don't see a way to form the product without explicitly computing the product of inverses.

twhunt committed:
matrices.py:

Introduced a new function _find_reasonable_pivot_naive() that
searches within a column for a suitable pivot.
This function has the same input and output signature as
_find_reasonable_pivot(), however it does not necessarily call
simplify(), which can be costly.

Refactored LUdecomposition_Simple() to use
_find_reasonable_pivot_naive().

twhunt committed:
Unit tests for matrices._find_reasonable_pivot_naive():

test_find_reasonable_pivot_naive_finds_guaranteed_nonzero1():
Tests if find_reasonable_pivot_naive() finds a guaranteed nonzero
pivot when one exists, and caller requests no simplifications be
attempted during the pivot search.

test_find_reasonable_pivot_naive_finds_guaranteed_nonzero2():
Tests if find_reasonable_pivot_naive() finds a guaranteed nonzero
pivot when one exists after some pivot candidates have been
simplified.

test_find_reasonable_pivot_naive_simplifies():
Tests if find_reasonable_pivot_naive() simplifies candidate pivots
and correctly reports their locations and simplifications when the
caller requests simplifications during the pivot search.

twhunt committed:
matrices.py:
Removed trailing whitespace.

twhunt commented:
@siefkenj I added a helper _find_reasonable_pivot_naive() that mimics the inputs and output of _find_reasonable_pivot().

twhunt commented:
Is there anything else I need to do to get this pull request reviewed and merged? Thanks!

siefkenj commented:
@twhunt I will review again this weekend.  Sorry for the delay.  I've been a bit busy!

twhunt commented:
@siefkenj I can sympathize. Thanks!

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -4150,35 +4252,129 @@ def LUdecomposition_Simple(self, iszerofunc=_iszero):
+                    lu[pivot_row + offset, pivot_col] = val
+
+                candidate_pivot_row = None if pivot_row_offset is None else pivot_row + pivot_row_offset
+                break
```
This is somehow confusing to me.  It would be more straight forward to say

```
while piv_col < self.cols:
   ...
   if pivot_value is not None:
       ....
       break
   pivot_cols += 1
```
and void the use of both a `continue` and a `break` in the same loop; especially a break that occurs at the end of the loop body.

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -4150,35 +4252,129 @@ def LUdecomposition_Simple(self, iszerofunc=_iszero):
+                # strictly less than min(num rows, num cols)
+                # Mimic behavior of previous implementation, by throwing a
+                # ValueError.
+
```
no blank line here

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -4150,35 +4252,129 @@ def LUdecomposition_Simple(self, iszerofunc=_iszero):
+        row_swaps = []
+
+        pivot_col = 0
+        for pivot_row in range(0, lu.rows-1):
```
`lu.rows - 1` instead of `lu.rows-1`

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -4150,35 +4252,129 @@ def LUdecomposition_Simple(self, iszerofunc=_iszero):
+                # Swap pivot row of U with candidate pivot row.
+                for col in range(pivot_col, lu.cols):
+                    lu[pivot_row, col], lu[candidate_pivot_row, col] =\
+                        lu[candidate_pivot_row, col], lu[pivot_row, col]
```
Is there a reason you're not using slices here?

twhunt committed:
sympy/matrices/matrices.py:

Tidied code.
Refactored pivot search while loop to break based on boolean
instead of break statement.
Added documentation and comments to _find_reasonable_pivot_naive().
In _find_reasonable_pivot_naive(), add multiple entries to
indeterminates even when at most one entry will be used.

jksuom commented:
There is one assertion in test_LUdecomp that should be modified: permute instead of permuteBkwd.

twhunt committed:
Merge branch 'master' of https://github.com/sympy/sympy

twhunt committed:
Merge branch 'master' into lu_decomp_extend

twhunt committed:
sympy/matrices/tests/test_matrices.py:

Replaced permuteBckwd(p) with permute_rows(p, 'backward').

twhunt commented:
Thanks! Hopefully, I merged from the upstream master branch correctly.

jksuom commented:
This has grown bigger than what I expected but it looks good on the whole.
Thanks to @siefkenj for the review. Anything else that you'd want to point out?

siefkenj commented:
@twhunt great job!  @jksuom looks ready for a merge to me!

jksuom commented:
This looks ready to me too. Thanks for you work, @twhunt .

twhunt commented:
@siefkenj @jksuom 
Thanks for your guidance, and the warm introduction to then open source world!
